src/tlo/methods/hsi_generic_first_appts.py

```diff
@@ -6,369 +6,179 @@
+            {}
+        )
+
+    def _diagnosis_fn(self, tests, use_dict: bool = False, report_tried: bool = False) -> Any:
+        """
+        Passed to modules when determining HSI_Events to be scheduled based on
+        this generic appointment. Intended as the diagnosis_fn argument to the
+        Module.do_at_generic_{non_}_emergency.
+
+        Class-level definition avoids the need to redefine this method each time
+        the .apply() method is called.
+
+        :param tests: The name of the test(s) to run via the diagnosis manager.
+        :param use_dict_for_single: If True, the return type will be a dictionary
+        even if only one test was requested.
+        :param report_dxtest_tried: Report if a test was attempted but could not
+        be carried out due to EG lack of consumables, etc.
+        :returns: Test results as dictionary key/value pairs.
+        """
+        return self.healthcare_system.dx_manager.run_dx_test(
+            tests,
+            hsi_event=self,
+            use_dict_for_single=use_dict,
+            report_dxtest_tried=report_tried,
+        )
+
+    def _do_on_generic_first_appt(self, squeeze_factor: float = 0.) -> None:
+        """
+        """
+        # Make top-level reads of information, to avoid repeat accesses.
+        modules: OrderedDict[str, "Module"] = self.sim.modules
+        schedule_hsi = self.healthcare_system.schedule_hsi_event
+        symptoms = modules["SymptomManager"].has_what(self.target)
 
-class HSI_GenericNonEmergencyFirstAppt(HSI_Event, IndividualScopeEventMixin):
-    """This is a Health System Interaction Event that represents the first interaction with the health system following
-     the onset of non-emergency symptom(s). This is the HSI that is generated by the HealthSeekingBehaviour module. By
-     default, it occurs at level '0' but it could occur also at other levels."""
+        # Dynamically create immutable container with the target's details stored.
+        # This will avoid repeat DataFrame reads when we call the module-level functions.
+        df = self.sim.population.props
+        patient_details = namedtuple("PatientDetails", df.columns)(*df.loc[self.target])
+
+        proposed_df_updates = {}
+
+        for module in modules.values():
+            event_info, df_updates = getattr(module, self.MODULE_METHOD_ON_APPLY)(
+                patient_id=self.target,
+                patient_details=patient_details,
+                symptoms=symptoms,
+                diagnosis_fn=self._diagnosis_fn,
+                consumables_checker=self.get_consumables,
+                facility_level=self.ACCEPTED_FACILITY_LEVEL,
+                treatment_id=self.TREATMENT_ID,
+                random_state=self.module.rng,
+            )
+            # Schedule any requested updates
+            for info in event_info:
+                event = info[0]
+                options = info[1]
+                schedule_hsi(event, **options)
```

Comment:
Is there an advantage / is it necessary to do the actual scheduling of the HSI events in this function rather than within the called `do_at_generic_first_appt` / `do_at_generic_first_emergency_appt` methods? It seems like having those methods return a list of HSI events and then schedule here ends up making a more complex return signature than needed, while as if we just had them directly call `schedule_hsi` (either by passing this in as an argument or calling out `HealthSystem.schedule_hsi_event` directly via `Module.sim.modules["HealthSystem"]`) we would avoid some of the boilerplate / complex return semantics we get from doing the scheduling indirectly.

Reply:
Only because the original issue suggested doing everything (DF updates + event scheduling) outside the module-level methods. Though scheduling the HSI events within the methods themselves presumably won't hurt - will do some reworking

Reply:
Ah sorry I guess you mean the last comment
>Alternatively at the cost of more complex function signatures, we could make do_at_generic_first_appt_non_emergency and do_at_generic_first_appt_emergency static methods of the modules, without direct access to the module attributes (including the population data frame). In this case we would need to pass through all objects that would be needed for dealing with scheduling following on HSI events as arguments.
Yeah I think thought at that point this could be an alternative to make things stricter in terms of what the `Module` implementations of the method could do, but specifically I was thinking here to force that they are static methods so they don't have access to the instance at all. Thanks for updating!